engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java

```diff
@@ -203,13 +205,22 @@ boolean hasWarnings(@CachedLibrary(limit = "3") WarningsLibrary warnings) {
   @ExportMessage
   Warning[] getWarnings(Node location, @CachedLibrary(limit = "3") WarningsLibrary warnings)
       throws UnsupportedMessageException {
-    ArrayRope<Warning> ropeOfWarnings = new ArrayRope<>();
+    if (cachedWarnings == null) {
```

Comment:
Note that if `cachedWarnings` would really be `@CompilationFinal`, you would have to do `CompilerDirectives.transferToInterpreter()` here. But that is just a side note. Better to remove `@CompilationFinal` annotations from the fields, as Jaroslav suggested.